mission/finite_state_machine/scripts/gate.py

```diff
@@ -7,39 +7,119 @@
+from vortex_msgs.srv import ControlMode, SetVelocity
 
 
 class GateSearch(smach.State):
     def __init__(self):
-        smach.State.__init__(self, outcomes=['preempted', 'succeeded', 'aborted'])           
+        smach.State.__init__(self, outcomes=['preempted', 'succeeded', 'aborted'])  
+
         self.landmarks_client = rospy.ServiceProxy('send_positions',request_position)
         rospy.wait_for_service('send_positions') 
         self.object = self.landmarks_client("gate").object
 
+        desired_velocity_topic = rospy.get_param("/controllers/velocity_controller/desired_velocity_topic")
+        self.velocity_ctrl_client = rospy.ServiceProxy(desired_velocity_topic,SetVelocity)
+        rospy.wait_for_service(desired_velocity_topic)
+
+        self.dp_pub = rospy.Publisher("/controllers/dp_data", DpSetpoint, queue_size=1)
+
         self.state_pub = rospy.Publisher('/fsm/state',String,queue_size=1)
 
         vtf_action_server = "/controllers/vtf_action_server"
         self.vtf_client = actionlib.SimpleActionClient(vtf_action_server, VtfPathFollowingAction)  
 
+        rospy.Subscriber("/odometry/filtered", Odometry, self.odom_cb)
+        self.odom = Odometry()
+
+    def odom_cb(self, msg):
+        self.odom = msg   
+
         
     def execute(self, userdata):
         self.state_pub.publish("gate_search")
+        
 
-        goal = VtfPathFollowingGoal()
-        goal.waypoints = [Point(5,0,-0.5)]
-        goal.forward_speed = 0.2
-        goal.heading = "path_dependent_heading"
-        self.vtf_client.wait_for_server()
-        self.vtf_client.send_goal(goal)
         rate = rospy.Rate(10)
-
         while not self.object.isDetected:
 
+            #SEARCH PATTERN
+            goal = VtfPathFollowingGoal()
+            goal.waypoints = [get_pose_in_front(self.odom.pose.pose,1).position]
+            goal.waypoints[0].z = - 1.1
+            goal.forward_speed = 0.2
+            goal.heading = "path_dependent_heading"
+            self.vtf_client.wait_for_server()
+            self.vtf_client.send_goal(goal)
+            while self.vtf_client.simple_state != actionlib.simple_action_client.SimpleGoalState.DONE and not self.object.isDetected:
+                self.object = self.landmarks_client("gate").object
+                print("SEARCHING FOR GATE ...")
+                print(self.object.objectPose.pose.position)
+                rate.sleep()
+            if self.object.isDetected:
+                break
+
+            goal = Pose()
+            goal.position = self.odom.pose.pose.position
+            goal.orientation = self.odom.pose.pose.orientation
+            goal = rotate_certain_angle(goal,45)
+            vel_goal = Twist()
+            vel_goal.angular.z = 0.2
+            vel_goal.linear.z = -0.01
+            vel_goal.linear.x = 0.01
+            self.velocity_ctrl_client(vel_goal,True)
+            rate = rospy.Rate(10)
```

Comment:
This piece of code appears multiple times. I suggest making it a function call instead. Also:
- rate only needs to be set once
- make magic constants tunable parameters